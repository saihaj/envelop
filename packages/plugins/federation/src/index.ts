import { Plugin } from '@envelop/types';
import { ApolloGateway, GatewayConfig } from '@apollo/gateway';
import { getOperationAST, print } from 'graphql';
import { InMemoryLRUCache, KeyValueCache } from 'apollo-server-caching';
import { GraphQLRequestMetrics, Logger } from 'apollo-server-types';

export interface FederationConfig {
  gateway?: GatewayConfig;
  load?: Parameters<ApolloGateway['load']>[0];
  metrics?: GraphQLRequestMetrics;
  cache?: KeyValueCache;
  logger?: Logger;
}

export const useFederation = (options: FederationConfig = {}): Plugin => {
  const {
    gateway: gatewayConfig = {},
    load: loadConfig = {},
    cache = new InMemoryLRUCache(),
    logger = console,
    metrics = {},
  } = options;
  const gateway = new ApolloGateway(gatewayConfig);
  gateway.load(loadConfig);
  let schemaHash: any;
  return {
    onPluginInit({ setSchema }) {
      if (gateway.schema) {
        setSchema(gateway.schema);
      }
    },
    onSchemaChange: ({ replaceSchema }) => {
      schemaHash = Date.now().toString();
      gateway.onSchemaChange(replaceSchema);
    },
    onExecute({ args, setExecuteFn }) {
      setExecuteFn(function federationExecutor() {
        const documentStr = print(args.document);
        const operation = getOperationAST(args.document, args.operationName ?? undefined);
        if (!operation) {
          throw new Error(`Operation ${args.operationName || ''} cannot be found in ${documentStr}`);
        }
        return gateway.executor({
          document: args.document,
          request: {
            query: documentStr,
            operationName: args.operationName ?? undefined,
            variables: args.variableValues ?? undefined,
          },
          operationName: args.operationName ?? null,
          cache,
          context: args.contextValue,
          queryHash: documentStr,
          logger,
          metrics,
          source: documentStr,
          operation,
          schema: args.schema,
          schemaHash,
        });
      });
    },
  };
};
