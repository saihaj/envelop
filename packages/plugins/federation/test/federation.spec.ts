import { LocalGraphQLDataSource } from '@apollo/gateway';
import { assertSingleExecutionValue, createTestkit } from '@envelop/testing';
import { execute } from 'graphql';
import { useFederation } from '../src';
import * as accounts from './fixtures/accounts';
import * as products from './fixtures/products';
import * as reviews from './fixtures/reviews';

describe('useFederation', () => {
  const query = /* GraphQL */ `
    # A query that the gateway resolves by calling all three services
    query GetCurrentUserReviews {
      me {
        username
        reviews {
          body
          product {
            name
            upc
          }
        }
      }
    }
  `;

  const useTestFederation = () =>
    useFederation({
      gateway: {
        localServiceList: [
          { name: 'accounts', typeDefs: accounts.typeDefs },
          { name: 'products', typeDefs: products.typeDefs },
          { name: 'reviews', typeDefs: reviews.typeDefs },
        ],
        buildService: definition => {
          switch (definition.name) {
            case 'accounts':
              return new LocalGraphQLDataSource(accounts.schema);
            case 'products':
              return new LocalGraphQLDataSource(products.schema);
            case 'reviews':
              return new LocalGraphQLDataSource(reviews.schema);
          }
          throw new Error(`Unknown service ${definition.name}`);
        },
      },
    });

  it('Should override execute function', async () => {
    const onExecuteSpy = jest.fn();

    const testInstance = createTestkit([
      useTestFederation(),
      {
        onExecute: onExecuteSpy,
      },
    ]);

    await testInstance.execute(query);

    expect(onExecuteSpy).toHaveBeenCalledTimes(1);
    expect(onExecuteSpy.mock.calls[0][0].executeFn).not.toBe(execute);
    expect(onExecuteSpy.mock.calls[0][0].executeFn.name).toBe('federationExecutor');
  });

  it('Should execute correctly', async () => {
    const testInstance = createTestkit([useTestFederation()]);
    const result = await testInstance.execute(query);
    assertSingleExecutionValue(result);
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });
});
