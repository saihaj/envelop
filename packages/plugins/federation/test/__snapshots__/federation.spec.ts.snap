// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFederation Should execute correctly 1`] = `
Object {
  "me": Object {
    "reviews": Array [
      Object {
        "body": "Love it!",
        "product": Object {
          "name": "Table",
          "upc": "1",
        },
      },
      Object {
        "body": "Too expensive.",
        "product": Object {
          "name": "Couch",
          "upc": "2",
        },
      },
    ],
    "username": "@ada",
  },
}
`;
